### PR TITLE
Various connectivity enhancements

### PR DESCRIPTION
**Pull request by zvea**
I made a bunch of changes this morning, mainly around:
- making it possible to connect to a Carvera over TCP instead of serial;
- making it possible to connect to the proxy from another machine (rather than only via loopback);
- making it possible to use the pendant even when there's no controller app connected;
- making things generally more configurable;
- tidying up small things here and there.

I realise these are quite significant changes.  Let me know if you're happy to keep receiving these or would prefer I just maintained my own fork.

### CODE REVIEW

**Comment by zvea** [comment]
That's a great point.  This implementation only sends its own commands when there is no controller app connected and so there are no comms to clobber.  When there is an app connected, the proxy is 100% passive.

File: index.ts [hunk 1]
```diff
@@ -181,4 +193,7 @@ async function main() {
     console.log('System online!');
 }
 
+if (!Config.DEBUG_LOGGING) {
+    console.debug = function() {}
```
Comment (Doridian) [inline index.ts:197]: Please don't overwrite default JS functions. I would just export a debug log function from a util file that depends on this variable set or not.

File: pendant/device.ts [hunk 2]
```diff
@@ -117,12 +117,14 @@ export class PendantDevice extends EventEmitter {
 
         for (const button of report.buttons) {
             if (!this.pressedButtons.has(button)) {
+                console.debug('button_down:', button);
```
Comment (Doridian) [inline pendant/device.ts:120]:
```suggestion
                console.debug('button_down', button);
```

File: pendant/device.ts [hunk 3]
```diff
@@ -117,12 +117,14 @@ export class PendantDevice extends EventEmitter {
...
                 this.emit('button_down', button);
             }
         }
 
         for (const oldButton of this.pressedButtons) {
             if (!report.buttons.has(oldButton)) {
+                console.debug('button_up:', oldButton);
```
Comment (Doridian) [inline pendant/device.ts:127]:
```suggestion
                console.debug('button_up', oldButton);
```

**Comment by Doridian** [comment]
I read through and left a couple comments. Will need to do a more thorough reading, this is the first pass and it looks good for the most part. Just some nitpicks mainly

**Comment by zvea** [comment]
Thanks for the review.  I've implemented proper logging, which I think should address your comments.

**Comment by Doridian** [comment]
Ended up making a couple of changes to your PR after testing it for my Carvera:
- 127.255.255.255 (the bcast address of loopback) does not work, the controller won't see it. So I hardcoded 127.0.0.1 for the loopback interface (the old behaviour)
- I made `PROXY_IP` default to `127.0.0.1` to provide a safe default for users to not accidently invite anyone on their network to connect to their CNC mill unless they explicitly wish to do so.

**Comment by zvea** [comment]
Thanks.  On the subject of config, I'm intending to make button functions configurable.  There are lots of unused buttons and a few things I'd like those buttons to do.  At the same time I recognise that different users might want to assign different functions to those buttons.

**Comment by Doridian** [comment]
@zvea Absolutely a great idea. Could make it some like JSON config and then in the future make a generator (like some tiny static website that you just click the buttons on, and assign what they're supposed to do or something).

And then maybe slowly adding more and more configurability, like setting what the step setting knobs do etc

**Pull request merged**